lib/webauthn/attestation_statement/android_safetynet.rb

```diff
@@ -0,0 +1,80 @@
+# frozen_string_literal: true
+
+require "jwt"
+require "openssl"
+require "webauthn/attestation_statement/base"
+
+module WebAuthn
+  # Implements https://www.w3.org/TR/2018/CR-webauthn-20180807/#android-safetynet-attestation
+  module AttestationStatement
+    class AndroidSafetynet < Base
+      def self.default_trust_store
+        OpenSSL::X509::Store.new.tap { |trust_store| trust_store.set_default_paths }
+      end
+
+      def valid?(authenticator_data, client_data_hash, trust_store: self.class.default_trust_store)
+        trusted_attestation_certificate?(trust_store) &&
+          valid_signature? &&
+          valid_attestation_domain? &&
+          valid_version? &&
+          valid_nonce?(authenticator_data, client_data_hash) &&
+          cts_profile_match?
+      end
+
+      private
+
+      def trusted_attestation_certificate?(trust_store)
+        signing_certificates.each do |certificate|
+          trust_store.add_cert(certificate)
+        end
+        trust_store.verify(attestation_certificate)
+      end
+
+      def valid_signature?
+        JWT.decode(statement["response"], attestation_certificate.public_key, true, algorithm: 'RS256')
+      rescue JWT::VerificationError
+        false
+      end
+
+      def valid_attestation_domain?
+        subject = attestation_certificate.subject.to_a
+        subject.assoc('CN')[1] == "attest.android.com"
+      end
+
+      def valid_version?
+        !statement["ver"].empty?
+      end
+
+      def valid_nonce?(authenticator_data, client_data_hash)
+        nonce = unverified_jws_result[0]["nonce"]
+        nonce == verification_data(authenticator_data, client_data_hash)
+      end
+
+      def cts_profile_match?
+        unverified_jws_result[0]["ctsProfileMatch"]
+      end
+
+      def verification_data(authenticator_data, client_data_hash)
+        Digest::SHA256.base64digest(authenticator_data.data + client_data_hash)
+      end
+
+      def attestation_certificate
+        attestation_certificate_chain[0]
+      end
+
+      def signing_certificates
+        attestation_certificate_chain[1..-1]
+      end
+
+      def attestation_certificate_chain
+        @attestation_certificate_chain ||= unverified_jws_result[1]["x5c"].map do |cert|
+          OpenSSL::X509::Certificate.new(Base64.strict_decode64(cert))
+        end
+      end
+
+      def unverified_jws_result
+        JWT.decode(statement["response"], nil, false)
```

Comment:
What do you think if we call `decode` just once the way we call it in `valid_signature?` and use the output of that as the `jws_result`? We could add a comment to that method that states the signature is validated as part of that process. We're parsing and decoding base64 segments more times than we need it to be done.

I'm still doubting we actually need the `jwt` dependency all together because we're actually using just the decode and verify flow of 1 of the 5 algorithms that this gem allows. However I acknowledge that it makes some checks in the decoding process that I would dislike to do here.

Reply:
I went through this again. I'm good with using `jwt`.
If we manage to have decode called only once for the `valid_signature?`, we're good to go.

I suggest calling `valid_signature?` as the first method in the `valid?`. And then:

```
def valid_signature?
  !!verified_jws
end

def verified_jws
  return @verified_jws unless @verified_jws.nil?

  @verified_jws =
    begin
      JWT.decode(statement["response"], attestation_certificate.public_key, true, algorithm: 'RS256')
    rescue JWT::VerificationError
      false
    end
end
```

WDTY?